### PR TITLE
Implement reverse 3PL and 3PU lookups

### PR DESCRIPTION
**Pull request by leonerd**

### CODE REVIEW

File: lib/bridge.js [hunk 1]
```diff
@@ -312,6 +312,28 @@ Bridge.prototype._customiseAppservice = function() {
 
         this.addAppServicePath({
             method: "GET",
+            path: "/_matrix/app/:ver/thirdparty/location",
+            handler: function(req, res) {
+                if (req.params.ver !== "unstable" || !lookupController.parseLocation) {
```
Comment (lukebarnard1) [inline lib/bridge.js]: This needs a comment I think, not obvious that ```parseLocation``` is a function.

File: lib/bridge.js [hunk 2]
```diff
@@ -312,6 +312,28 @@ Bridge.prototype._customiseAppservice = function() {
...
+                    res.status(404).send("Cannot GET " + req.path + "\r\n");
```
Comment (lukebarnard1) [inline lib/bridge.js]: This error message isn't very useful

File: lib/bridge.js [hunk 3]
```diff
@@ -332,6 +354,28 @@ Bridge.prototype._customiseAppservice = function() {
                 );
             }
         });
+
+        this.addAppServicePath({
+            method: "GET",
+            path: "/_matrix/app/:ver/thirdparty/user",
+            handler: function(req, res) {
+                if (req.params.ver !== "unstable" || !lookupController.parseUser) {
+                    res.status(404).send("Cannot GET " + req.path + "\r\n");
```
Comment (lukebarnard1) [inline lib/bridge.js]: As above

File: lib/bridge.js (hunk 2 above)
Comment (leonerd) [inline lib/bridge.js]: Ah this one. @Kegsay commented on that too in the original functions.
That is attempting to preserve the same error message that `express` generates if it receives a request to a path it has no routes for.

Comment (lukebarnard1) [inline lib/bridge.js]: But it's still not useful, someone might misinterpret that as "GET" is not valid or any part of the path  is invalid. Why not, "version not supported" or "function not supported" (for the ```!lookupController.parseLocation``` bit.

Comment (leonerd) [inline lib/bridge.js]: Hmyes; perhaps the version check should be separated out into a different return value. I'll pop that (into all five)

**Comment by leonerd** [comment]
PTAL - made something of a reworking of it. Each `app.get()` route is now only installed if there's a controller function for it anyway, so we don't have to test and return the 404 any more.
Additionally, I moved all the 3PE lookup logic into its own function so as to reduce the ever-growing indentation level here.

File: lib/bridge.js (hunk 2 above)
Comment (leonerd) [inline lib/bridge.js]: Now done

File: lib/bridge.js (hunk 1 above)
Comment (leonerd) [inline lib/bridge.js]: Hopefully now addressed by indirecting it via a variable called `parseLocationFunc`

**Pull request merged**